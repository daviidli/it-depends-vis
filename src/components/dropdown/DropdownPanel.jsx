import React from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Select } from 'antd';
import {
	pipe, dropWhile, dropLastWhile, init, tail, map, addIndex
} from 'ramda';

import './DropdownPanel.scss';

const createOptionItem = (value, i) => <Select.Option key={`option-${i}`} value={value}>{ value }</Select.Option>;
const notEquals = x => y => x !== y;
const mappedIndex = addIndex(map);

const createStartCommitOptions = commit => pipe(
	dropLastWhile(notEquals(commit)),
	init,
	mappedIndex(createOptionItem)
);
const createEndCommitOptions = commit => pipe(
	dropWhile(notEquals(commit)),
	tail,
	mappedIndex(createOptionItem)
);

const DropdownPanel = ({
	startCommit, endCommit, setStartCommit, setEndCommit, commits
}) => (
	<Row>
		<Col span={24} className="dropdownCol">
			<div>Staring Commit</div>
			<Select
				className="dropdownSelect"
				size="middle"
				value={startCommit}
				onSelect={setStartCommit}
			>
				{ createStartCommitOptions(endCommit)(commits) }
			</Select>
		</Col>
		<Col span={24} className="dropdownCol">
			<div>Ending Commit</div>
			<Select
				className="dropdownSelect"
				size="middle"
				value={endCommit}
				onSelect={setEndCommit}
			>
				{ createEndCommitOptions(startCommit)(commits) }
			</Select>
		</Col>
	</Row>
);

DropdownPanel.propTypes = {
	startCommit: PropTypes.string.isRequired,
	endCommit: PropTypes.string.isRequired,
	setStartCommit: PropTypes.func.isRequired,
	setEndCommit: PropTypes.func.isRequired,
	commits: PropTypes.arrayOf(PropTypes.string).isRequired
};

export default DropdownPanel;
