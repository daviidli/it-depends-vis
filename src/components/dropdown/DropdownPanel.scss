.dropdownSelect {
	width: 100% !important;
}

.dropdownCol {
	padding: 1em;
}
