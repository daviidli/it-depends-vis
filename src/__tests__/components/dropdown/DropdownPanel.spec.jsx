import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Select } from 'antd';
import { clone } from 'ramda';
import DropdownPanel from '../../../components/dropdown/DropdownPanel';

const defaultProps = {
	startCommit: 0,
	endCommit: 5,
	setStartCommit: jest.fn(),
	setEndCommit: jest.fn(),
	commits: ['a', 'b', 'c', 'd', 'e', 'f']
};

const setup = (props = defaultProps) => {
	const wrapper = shallow(<DropdownPanel {...props} />);
	return {
		wrapper,
		select: wrapper.find(Select)
	};
};

describe('DropdownPanel component', () => {
	it('should match snapshot', () => {
		const { wrapper } = setup();
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should match snapshot with correct options', () => {
		const props = clone(defaultProps);
		props.startCommit = 1;
		props.endCommit = 3;
		const { wrapper } = setup(props);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it('should call setStartCommit on select', () => {
		const mockSetStartCommit = jest.fn();
		const props = clone(defaultProps);
		props.setStartCommit = mockSetStartCommit;
		const { select } = setup(props);
		select.at(0).simulate('select', 'b');
		expect(mockSetStartCommit).toHaveBeenCalledTimes(1);
		expect(mockSetStartCommit).toHaveBeenCalledWith(1);
	});

	it('should call setEndCommit on select', () => {
		const mockSetEndCommit = jest.fn();
		const props = clone(defaultProps);
		props.setEndCommit = mockSetEndCommit;
		const { select } = setup(props);
		select.at(1).simulate('select', 'd');
		expect(mockSetEndCommit).toHaveBeenCalledTimes(1);
		expect(mockSetEndCommit).toHaveBeenCalledWith(3);
	});
});
