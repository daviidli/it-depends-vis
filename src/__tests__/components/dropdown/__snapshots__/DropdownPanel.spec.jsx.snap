// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownPanel component should match snapshot 1`] = `
<Row
  gutter={0}
>
  <Col
    className="dropdownCol"
    span={24}
  >
    <div>
      Staring Commit
    </div>
    <Select
      bordered={true}
      choiceTransitionName="zoom"
      className="dropdownSelect"
      onSelect={[MockFunction]}
      size="middle"
      transitionName="slide-up"
      value="a"
    >
      <Option
        key="option-0"
        value="a"
      >
        a
      </Option>
      <Option
        key="option-1"
        value="b"
      >
        b
      </Option>
      <Option
        key="option-2"
        value="c"
      >
        c
      </Option>
      <Option
        key="option-3"
        value="d"
      >
        d
      </Option>
      <Option
        key="option-4"
        value="e"
      >
        e
      </Option>
    </Select>
  </Col>
  <Col
    className="dropdownCol"
    span={24}
  >
    <div>
      Ending Commit
    </div>
    <Select
      bordered={true}
      choiceTransitionName="zoom"
      className="dropdownSelect"
      onSelect={[MockFunction]}
      size="middle"
      transitionName="slide-up"
      value="f"
    >
      <Option
        key="option-0"
        value="b"
      >
        b
      </Option>
      <Option
        key="option-1"
        value="c"
      >
        c
      </Option>
      <Option
        key="option-2"
        value="d"
      >
        d
      </Option>
      <Option
        key="option-3"
        value="e"
      >
        e
      </Option>
      <Option
        key="option-4"
        value="f"
      >
        f
      </Option>
    </Select>
  </Col>
</Row>
`;

exports[`DropdownPanel component should match snapshot with correct options 1`] = `
<Row
  gutter={0}
>
  <Col
    className="dropdownCol"
    span={24}
  >
    <div>
      Staring Commit
    </div>
    <Select
      bordered={true}
      choiceTransitionName="zoom"
      className="dropdownSelect"
      onSelect={[MockFunction]}
      size="middle"
      transitionName="slide-up"
      value="b"
    >
      <Option
        key="option-0"
        value="a"
      >
        a
      </Option>
      <Option
        key="option-1"
        value="b"
      >
        b
      </Option>
      <Option
        key="option-2"
        value="c"
      >
        c
      </Option>
    </Select>
  </Col>
  <Col
    className="dropdownCol"
    span={24}
  >
    <div>
      Ending Commit
    </div>
    <Select
      bordered={true}
      choiceTransitionName="zoom"
      className="dropdownSelect"
      onSelect={[MockFunction]}
      size="middle"
      transitionName="slide-up"
      value="d"
    >
      <Option
        key="option-0"
        value="c"
      >
        c
      </Option>
      <Option
        key="option-1"
        value="d"
      >
        d
      </Option>
      <Option
        key="option-2"
        value="e"
      >
        e
      </Option>
      <Option
        key="option-3"
        value="f"
      >
        f
      </Option>
    </Select>
  </Col>
</Row>
`;
